### PR TITLE
adding some fixes to help on azure request throlling

### PR DESCRIPTION
**Pull request by maksd**
**Description**

Changes proposed in this pull request:

- adding Zipkin tracing to help find issues
- forcing Azure clients to not retry on 429 errors (TooManyRequests), because by default it retry 3 times
- remove unnecessary api calls
- adding some backoff logic on 429 retry to help on API rate limits
- typo fixes
- other small changes

### CODE REVIEW

File: components/metris/internal/provider/azure/clients.go [hunk 1]
```diff
@@ -195,18 +197,30 @@ func (c *client) GetResourceGroup(ctx context.Context, name, filter string, logg
...
+	baseclient := *c.computeBaseClient
+	baseclient.RequestInspector = LogRequest(logger)
+	baseclient.ResponseInspector = LogResponse(logger)
+	skuClient := compute.ResourceSkusClient{BaseClient: baseclient}
+
+	// Only **location** filter is supported currently. :(
 	for skuList, err = skuClient.ListComplete(ctx, filter); skuList.NotDone(); err = skuList.NextWithContext(ctx) {
 		if err != nil {
-			return result, err
+			break
```
Comment (marcobebway) [inline components/metris/internal/provider/azure/clients.go:223]: Why do we need this change?

File: components/metris/internal/provider/azure/clients.go [hunk 2]
```diff
@@ -219,18 +233,29 @@ func (c *client) GetVMResourceSkus(ctx context.Context, filter string) (result [
...
+
+	baseclient := *c.computeBaseClient
+	baseclient.RequestInspector = LogRequest(logger)
+	baseclient.ResponseInspector = LogResponse(logger)
+	vmClient := compute.VirtualMachinesClient{BaseClient: baseclient}
 
 	for vmList, err = vmClient.ListComplete(ctx, rgname); vmList.NotDone(); err = vmList.NextWithContext(ctx) {
 		if err != nil {
-			return result, err
+			break
```
Comment (marcobebway) [inline components/metris/internal/provider/azure/clients.go:258]: Why do we need this change?

File: components/metris/internal/provider/azure/clients.go [hunk 3]
```diff
@@ -240,18 +265,29 @@ func (c *client) GetVirtualMachines(ctx context.Context, rgname string) (result
...
+
+	baseclient := *c.computeBaseClient
+	baseclient.RequestInspector = LogRequest(logger)
+	baseclient.ResponseInspector = LogResponse(logger)
+	diskClient := compute.DisksClient{BaseClient: baseclient}
 
 	for diskList, err = diskClient.ListByResourceGroupComplete(ctx, rgname); diskList.NotDone(); err = diskList.NextWithContext(ctx) {
 		if err != nil {
-			return result, err
+			break
```
Comment (marcobebway) [inline components/metris/internal/provider/azure/clients.go:290]: Why do we need this change?

File: components/metris/internal/provider/azure/clients.go [hunk 4]
```diff
@@ -264,18 +300,29 @@ func (c *client) GetDisks(ctx context.Context, rgname string) (result []compute.
...
+
+	baseclient := *c.networkBaseClient
+	baseclient.RequestInspector = LogRequest(logger)
+	baseclient.ResponseInspector = LogResponse(logger)
+	loadBalancersClient := network.LoadBalancersClient{BaseClient: baseclient}
 
 	for lbList, err = loadBalancersClient.ListComplete(ctx, rgname); lbList.NotDone(); err = lbList.NextWithContext(ctx) {
 		if err != nil {
-			return result, err
+			break
```
Comment (marcobebway) [inline components/metris/internal/provider/azure/clients.go:325]: Why do we need this change?

File: components/metris/internal/provider/azure/azure.go [hunk 5]
```diff
@@ -28,6 +34,13 @@ var (
 
 // NewAzureProvider returns a new Azure provider.
 func NewAzureProvider(config *provider.Config) provider.Provider {
+	// enable azure go-autorest tracing
+	if tracing.IsEnabled() {
+		if err := opencensus.Enable(); err != nil {
+			panic(err)
```
Comment (marcobebway) [inline components/metris/internal/provider/azure/azure.go:40]: Does it make sense to use `log.Panic(err)`?

File: components/metris/internal/provider/azure/azure.go [hunk 6]
```diff
@@ -144,70 +153,72 @@ func (a *Azure) clusterHandler(parentctx context.Context) {
...
-					logger.Warnf("could not find event hub resource groups, cluster may not be ready, retrying in %s: %s", a.config.PollInterval, err)
+			if instance.clusterResourceGroupName == "" {
+				// check if resource group exists in azure, if not, cluster may not be ready yet.
+				if _, err := instance.client.GetResourceGroup(parentctx, cluster.TechnicalID, "", logger); err != nil {
+					logger.With("error", err).Warnf("could not find resource group \"%s\", cluster may not be ready, retrying in %s", cluster.TechnicalID, a.config.PollInterval)
 					time.AfterFunc(a.config.PollInterval, func() { a.config.ClusterChannel <- cluster })
 
 					continue
+				} else {
+					instance.clusterResourceGroupName = cluster.TechnicalID
```
[the remark this answers is not in the document]
Comment (maksd) [reply components/metris/internal/provider/azure/azure.go:192]: you are right, I will remove it.

File: components/metris/internal/provider/azure/azure.go (hunk 5 above)
Comment (maksd) [reply components/metris/internal/provider/azure/azure.go:40]: no it does not, I will change that.

File: components/metris/internal/provider/azure/azure.go [hunk 7]
```diff
@@ -258,34 +272,98 @@ func (a *Azure) gatherMetrics(parentctx context.Context, workerlogger log.Logger
...
+		MaxIncomingBytesPT1M: 0,
+		MaxOutgoingBytesPT1M: 0,
+		IncomingRequestsPT5M: 0,
+		MaxIncomingBytesPT5M: 0,
+		MaxOutgoingBytesPT5M: 0,
+	}
+
+	if err == nil && len(instance.eventHubResourceGroupName) > 0 {
+		eventData.ResourceGroups = append(eventData.ResourceGroups, instance.eventHubResourceGroupName)
+		if eventhubData, err = instance.getEventHubMetrics(ctx, a.config.PollInterval, workerlogger); err == nil {
```
[the remark this answers is not in the document]
Comment (maksd) [reply components/metris/internal/provider/azure/azure.go:310]: yes, will fix it

**Review by marcobebway — APPROVED** [review]
/lgtm

**Review by franpog859 — APPROVED** [review]

**Pull request merged**